conn.go

```diff
@@ -0,0 +1,214 @@
+package coap
+
+import (
+	"net"
+	"sync/atomic"
+	"time"
+)
+
+type writeReq interface {
+}
+
+type writeReqBase struct {
+	req      Message
+	respChan chan error
+}
+
+type writeReqTCP struct {
+	writeReqBase
+}
+
+type writeReqUDP struct {
+	writeReqBase
+	sessionData *SessionUDPData
+}
+
+type conn interface {
+	LocalAddr() net.Addr
+	RemoteAddr() net.Addr
+	Write(w writeReq) error
+	Close() error
+}
+
+type connBase struct {
+	writeChan chan writeReq
+	closeChan chan bool
+	finChan   chan bool
+	closed    int32
+}
+
+func (conn *connBase) finishWrite() {
+	if !atomic.CompareAndSwapInt32(&conn.closed, conn.closed, 1) {
+		return
+	}
+	conn.closeChan <- true
+	<-conn.finChan
+}
+
+type connTCP struct {
+	connBase
+	connection net.Conn // i/o connection if TCP was used
+	num        int32
+}
+
+func (conn *connTCP) LocalAddr() net.Addr {
+	return conn.connection.LocalAddr()
+}
+
+func (conn *connTCP) RemoteAddr() net.Addr {
+	return conn.connection.RemoteAddr()
+}
+
+func (conn *connTCP) Close() error {
+	conn.finishWrite()
+	return conn.connection.Close()
+}
+
+func (conn *connTCP) Write(w writeReq) error {
+	if atomic.LoadInt32(&conn.closed) > 0 {
+		return ErrConnectionClosed
+	}
+	conn.writeChan <- w
```

Comment:
again consider timeout. i'm pointing this out not as a must but consideration for you with a question. what would happen if this blocks forever

Reply:
added timeout as arg. Fixed